File: ui/src/app/feature.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Feature } from './Product';

@Injectable({
  providedIn: 'root'
})
export class FeatureService {
  private baseUrl = 'http://localhost:8080/api/features';

  constructor(private http: HttpClient) { }
  internalName: string= "";

  addFeatureToProduct(productInternalName: string, feature: Feature): Observable<void> {
    const url = `${this.baseUrl}/${productInternalName}`;
    return this.http.post<void>(url, feature);
  }

  getParametersByFeatureInternalName(featureInternalName: string): Observable<Feature[]> {
    const url = `${this.baseUrl}/${featureInternalName}/parameters`;
    return this.http.get<Feature[]>(url);
  }

  getAllFeatures(): Observable<Feature[]> {
    return this.http.get<Feature[]>(this.baseUrl);
  }

  getFeatureByInternalName(featureInternalName: string): Observable<Feature> {
    const url = `${this.baseUrl}/${featureInternalName}`;
    return this.http.get<Feature>(url);
  }

  getFeaturesByProductInternalName(productInternalName: string): Observable<Feature[]> {
    const url = `${this.baseUrl}/product/${productInternalName}`;
    return this.http.get<Feature[]>(url);
  }

  updateFeature(feature: Feature): Observable<Feature> {
    const url = `${this.baseUrl}/update`;
    return this.http.put<Feature>(url, feature);
  }

  deleteFeature(feature: Feature): Observable<void> {
    console.log(feature.internalName);
    const url = `${this.baseUrl}?internalName=${feature.internalName}`;
    return this.http.delete<void>(url);
  }

  saveFeatureName(internalName: string): void{
    this.internalName = internalName;
  }

  getFeatureName(): string{
    return this.internalName;
  }
}

File: bakend/src/main/java/com/learnings/capstone/service/FeatureService.java
package com.learnings.capstone.service;

import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.learnings.capstone.dto.FeatureDTO;
import com.learnings.capstone.dto.ParameterDTO;
import com.learnings.capstone.entity.Feature;
import com.learnings.capstone.entity.Parameter;
import com.learnings.capstone.entity.Product;
import com.learnings.capstone.repository.FeatureRepository;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class FeatureService {

    @Autowired
    private FeatureRepository featureRepository;

    @Autowired
    private ProductService productService; 

    public void addFeatureToProduct(String productInternalName, FeatureDTO featureDTO) {
        Product product = productService.getProductEntityByInternalName(productInternalName);

        Feature feature = convertDTOToEntity(featureDTO);
        feature.setProduct(product);
        
        featureRepository.save(feature);
    }

    
    public FeatureDTO getFeatureByInternalName(String featureInternalName) {
        Feature feature = getFeatureEntityByInternalName(featureInternalName);
        return convertEntityToDTO(feature);
    }
    
    
    
    public List<FeatureDTO> getAllFeatures() {
        List<Feature> features = featureRepository.findAll();
        return features.stream()
        .map(this::convertEntityToDTO)
        .toList();
    }

    public FeatureDTO updateFeature(FeatureDTO featureDTO) {
        Feature existingFeature = featureRepository.getFeatureEntityByInternalName(featureDTO.getInternalName());
        updateFeatureFromDTO(existingFeature, featureDTO);
        featureRepository.save(existingFeature);
        return convertEntityToDTO(existingFeature);
    }
    
    public void deleteFeature(String internalName) {
        Feature existingFeature = featureRepository.getFeatureEntityByInternalName(internalName);
        featureRepository.delete(existingFeature);
    }
    
    public Feature convertDTOToEntity(FeatureDTO featureDTO) {
        Feature feature = new Feature();
        BeanUtils.copyProperties(featureDTO, feature);
        return feature;
    }
    
    public FeatureDTO convertEntityToDTO(Feature feature) {
        FeatureDTO featureDTO = new FeatureDTO();
        BeanUtils.copyProperties(feature, featureDTO);
        return featureDTO;
    }
    
    public void updateFeatureFromDTO(Feature feature, FeatureDTO featureDTO) {
        BeanUtils.copyProperties(featureDTO, feature);
    }
    
    
    public List<ParameterDTO> getParametersByFeatureInternalName(String featureInternalName) {
        Feature feature = getFeatureEntityByInternalName(featureInternalName);
        return feature.getParameters().stream()
        .map(this::convertParameterEntityToDTO)
        .collect(Collectors.toList());
    }
    
    
    public Feature getFeatureEntityByInternalName(String featureInternalName) {
        return featureRepository.getFeatureEntityByInternalName(featureInternalName);
    }
    
    private ParameterDTO convertParameterEntityToDTO(Parameter parameter) {
        ParameterDTO parameterDTO = new ParameterDTO();
        BeanUtils.copyProperties(parameter, parameterDTO);
        return parameterDTO;
    }
    
    public List<FeatureDTO> getFeaturesByProductInternalName(String productInternalName) {
        List<Feature> features = featureRepository.findByProduct_InternalName(productInternalName);
        return features.stream()
            .map(this::convertEntityToDTO)
            .toList();
    }
}
